File: Assignment Lab/kmp_algo.py

def calTempArray():
    arr = [0] * len(str2)
    j=0
    i=1
    while(i < len(str2)):
        if str2[j] == str2[i]:
            arr[i] = j + 1
            i+=1
            j+=1
        else:
            if j!=0:
                j = arr[j-1]
            else:
                i+=1
    
    return arr

def KMPAlgo():
    arr = calTempArray()
    i,j = 0,0
    while i < len(str1) and j < len(str2):
        if str1[i] == str2[j]:
            if j==len(str2) - 1:
                return i - j
            i+=1
            j+=1
        else:
            if j!=0:
                j = arr[j-1]
            else:
                i+=1

    return -1

str1 = input("Enter the main string: ")
str2 = input("Enter the pattern string: ")
print("Pattern exist at index:",KMPAlgo()) if KMPAlgo() >= 0 else print("Pattern doesn't exist...")

File: my_programs/kmp_algo.c
#include<stdio.h>
#include<string.h>

int *calcPatternArray(char str[]){
    static int arr[100];
    int j = 0,i = 1;

    int len = strlen(str) - 1;
    while(i < len){
        if(str[j] == str[i]){
            arr[i] = j + 1;
            j++;
            i++;
        }else{
            if(j != 0)
                j = arr[j - 1];
            else
                i++;
        }
    }
    return arr;
}

int kmpAlgo(char str1[],char str2[],int len1,int len2){
    int *patternArray = calcPatternArray(str2);
    int i = 0,j = 0;
    while(i < len1 && j < len2){
        if(str1[i] == str2[j]){
            if( j == len2 - 1)
                return i - j;
            i++;
            j++;
        } else {
            if(j != 0)
                j = patternArray[j - 1];
            else
                i++;
        }
    }
    return -1;
}

int main(){
    char str1[100],str2[100];
    int i;

    printf("Enter the main string...\n");
    fgets(str1,sizeof(str1),stdin);
    printf("Enter the pattern string...\n");
    fgets(str2,sizeof(str2),stdin);

    int len1 = strlen(str1) - 1;
    int len2 = strlen(str2) - 1;

    int index = kmpAlgo(str1,str2,len1,len2);
    if(index >= 0)
        printf("pattern exist at %d",index);
    else
        printf("pattern doesn't exist");   
}